File: tiledb/sys2/cpp/mode.cc
#include <string>

#include <tiledb/tiledb.h>

#include "exception.h"
#include "tiledb-sys2/src/mode.rs.h"

namespace tiledb::rs {

tiledb_query_type_t to_cpp_mode(Mode mode) {
  switch (mode) {
    case Mode::Read:
      return TILEDB_READ;
    case Mode::Write:
      return TILEDB_WRITE;
    case Mode::Delete:
      return TILEDB_DELETE;
    case Mode::Update:
      return TILEDB_UPDATE;
    case Mode::ModifyExclusive:
      return TILEDB_MODIFY_EXCLUSIVE;
    default:
      throw TileDBError("Invalid TileOrder for conversion.");
  }
}

Mode to_rs_mode(tiledb_query_type_t mode) {
  switch (mode) {
    case TILEDB_READ:
      return Mode::Read;
    case TILEDB_WRITE:
      return Mode::Write;
    case TILEDB_DELETE:
      return Mode::Delete;
    case TILEDB_UPDATE:
      return Mode::Update;
    case TILEDB_MODIFY_EXCLUSIVE:
      return Mode::ModifyExclusive;
    default:
      throw TileDBError("Invalid tiledb_layout_t for TileOrder conversion.");
  }
}

}  // namespace tiledb::rs

File: tiledb/sys2/src/mode.rs
use tiledb_common::array::Mode;

use crate::error::TryFromFFIError;

#[cxx::bridge(namespace = "tiledb::rs")]
mod ffi {
    #[derive(Debug)]
    pub enum Mode {
        Read,
        Write,
        Delete,
        Update,
        ModifyExclusive,
    }
}

pub use ffi::Mode as FFIMode;

impl From<Mode> for FFIMode {
    fn from(mode: Mode) -> FFIMode {
        match mode {
            Mode::Read => FFIMode::Read,
            Mode::Write => FFIMode::Write,
            Mode::Delete => FFIMode::Delete,
            Mode::Update => FFIMode::Update,
            Mode::ModifyExclusive => FFIMode::ModifyExclusive,
        }
    }
}

impl TryFrom<FFIMode> for Mode {
    type Error = TryFromFFIError;

    fn try_from(mode: FFIMode) -> Result<Self, Self::Error> {
        let mode = match mode {
            FFIMode::Read => Mode::Read,
            FFIMode::Write => Mode::Write,
            FFIMode::Delete => Mode::Delete,
            FFIMode::Update => Mode::Update,
            FFIMode::ModifyExclusive => Mode::ModifyExclusive,
            _ => return Err(TryFromFFIError::from_mode(mode)),
        };
        Ok(mode)
    }
}
